### PR TITLE
Actionsets: Add ability for plugins to add actions for core actionsets (i.e. `folders:edit`)

### PR DESCRIPTION
**Pull request by eleijonmarck**
**why**
we need a way for plugins to register they actionsets to have roles they pick from for users to have access to specific plugins.

**what**
k6 / SLO teams have requested for us to be able to register actions as part of their registration. This enables a way to specify a set of actions that would then be associated (in-memory) for the `folders:edit` for instance to be evaluated together with the additional actions.

```json
# plugins.json
...
    "actionsets": [
      {
        "action": "folders:edit",
        "actions": ["k6app:tests.read", "k6app:tests.write"]
      }
    ],
  "iam": {
    "permissions": [
      {
        "action": "users.permissions:read",
        "scope": "users:*"
      }
    ]
  },
...
```

**implementation details**
- adds declare action set as a plugin Registry (same as role registry)
- modifys the existing action sets to add on top of folders or dashboard actions sets 

**Out of scope**
- creation of action sets 
- ability to remove the added actions from the plugin 

---

how to test to add actionsets to plugins:
- clone the [grafana-plugin-examples](https://github.com/grafana/grafana-plugin-examples)
- build the assets for the plugin
```bash
mage -v (building the plugin assets)
yarn install
yarn build (building frontend assets)
```
- edit the plugins.json with actionsets
- symlink plugin with app-with-rbac to grafana
```bash
pwd
~/dev/grafana
cd data/plugins/
ln -sfn ~/dev/grafana/app-with-rbac/dist grafana-appwithrbac-app
```

[implementation doc with notes](https://docs.google.com/document/d/1BjMaluG-OdFb5_W34CgKyUHuLBq5nyoJUziMnxwvb0k/edit)

image

[actionsets registration proposals or implementation strategies](https://docs.google.com/document/d/1L1-PeuwwnKIH8Y5km3O3gEuDUcIr2eiBsBGx5TMtIGE/edit)

https://github.com/grafana/identity-access-team/issues/615

### CODE REVIEW

File: pkg/services/serviceaccounts/models.go [hunk 1]
```diff
@@ -18,7 +17,10 @@ var (
 const (
 	ServiceAccountPrefix = "sa-"
 	ExtSvcPrefix         = "extsvc-"
-	ExtSvcLoginPrefix    = ServiceAccountPrefix + extsvcauth.TmpOrgIDStr + "-" + ExtSvcPrefix
+
+	// TmpOrgID is the orgID we use while global service accounts are not supported.
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/models.go:21]: needed to resolve cyclic dependancy

File: pkg/services/sqlstore/migrations/cloud_migrations.go [hunk 2]
```diff
@@ -120,5 +120,4 @@ func addCloudMigrationsMigrations(mg *Migrator) {
 		"updated":     "updated",
 		"finished":    "finished",
 	})
-
```
Comment (eleijonmarck) [inline pkg/services/sqlstore/migrations/cloud_migrations.go:123]: fix that is not part of the pr

File: pkg/services/serviceaccounts/extsvcaccounts/service.go [hunk 3]
```diff
@@ -152,6 +161,24 @@ func (esa *ExtSvcAccountsService) SaveExternalService(ctx context.Context, cmd *
...
+		svc := *esa.actionSetsSvc
+		// Add the new action sets
+		if cmd.Self.ActionSets != nil {
+			for _, actionSet := range cmd.Self.ActionSets {
+				// TODO: need to know which resource, permissions and actions are needed
+				svc.StoreActionSet(slug, actionSet.Action, actionSet.Actions)
+				// add the actionset as part of the permissions
+				// convert actionset to permissions
+				// add the permissions to the service account
+				cmd.Self.Permissions = append(cmd.Self.Permissions, ac.Permission{
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/extsvcaccounts/service.go:175]: this stores the permission into the db.

this could be removed from this PR if needed

File: pkg/services/pluginsintegration/serviceregistration/serviceregistration.go [hunk 4]
```diff
@@ -8,13 +8,15 @@ import (
...
 	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/accesscontrol/resourcepermissions"
 	"github.com/grafana/grafana/pkg/services/extsvcauth"
 	"github.com/grafana/grafana/pkg/services/featuremgmt"
 	"github.com/grafana/grafana/pkg/services/pluginsintegration/pluginsettings"
 )
 
 type Service struct {
 	featureEnabled bool
+	features       featuremgmt.FeatureToggles
```
Comment (wbrowne) [inline pkg/services/pluginsintegration/serviceregistration/serviceregistration.go:19]: Nit: Doesn't look like this is used

**Comment by marefr** [comment]
Would it make sense to prefix/enforce plugin ID of action name in action sets to enforce uniqueness per plugin or is that already handled elsewhere perhaps or how do you know that an action belongs to a certain plugin for example?

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

**Comment by gamab** [comment]
The iam section contains the permissions for the plugin itself, roles are defined outside of this section, so should actionSets.

```json
 "actionsets": [
      {
        "action": "test.writer",
        "actions": ["users.permissions:read", "dashboards:read"]
      }
    ],
  "iam": {
    "permissions": [
      {
        "action": "users.permissions:read",
        "scope": "users:*"
      }
    ]
  },
```

**Comment by marefr** [comment]
> image
> ```json
> # plugins.json
> ...
>   "iam": {
>     "actionsets": [
>       {
>         "action": "test.writer",
>         "actions": ["users.permissions:read", "dashboards:read"]
>       }
>     ],
>     "permissions": [
>       {
>         "action": "users.permissions:read",
>         "scope": "users:*"
>       }
>     ]
>   },
> ...
> ```

@gamab seems like action set actions are stored in the database? That's why I asked/was curious whether that's a potential problem if they're not "namespaced/prefixed" with plugin id to make them unique

**Comment by gamab** [comment]
@marefr your point makes perfect sense, I too think they should be prefixed by the plugin ID. 

I was just challenging where the action sets should be defined, let me fix the comment's example.

**Comment by eleijonmarck** [comment]
@gamab @marefr are you seeing that we would enforce the pluginid inside of the actionset names in the json?

Or are you seeing that we would store the actionset prefixed with the pluginid to be able to identify actionsets by plugin.

I am assuming you mean the later

**Comment by marefr** [comment]
> I am assuming you mean the later

That was my idea

File: pkg/services/accesscontrol/resourcepermissions/store.go [hunk 5]
```diff
@@ -824,15 +831,6 @@ func (s *InMemoryActionSets) ExpandActionSetsWithFilter(permissions []accesscont
 	return expandedPermissions
 }
 
-// GetActionSet returns the action set for the given action.
-func (s *InMemoryActionSets) GetActionSet(actionName string) []string {
```
Comment (eleijonmarck) [inline pkg/services/accesscontrol/resourcepermissions/store.go:828]: unused

File: pkg/services/pluginsintegration/pipeline/steps.go [hunk 6]
```diff
@@ -94,6 +94,34 @@ func (r *RegisterPluginRoles) Register(ctx context.Context, p *plugins.Plugin) (
...
+		log:          log.New("plugins.actionsets.registration"),
+		roleRegistry: registry,
+	}
+}
+
+// Register registers the plugin roles with the role registry.
+func (r *RegisterActionSets) Register(ctx context.Context, p *plugins.Plugin) (*plugins.Plugin, error) {
+	r.log.Info("Registering actionsets", "pluginId", p.ID, "actionSets", p.ActionSets)
+	if err := r.roleRegistry.DeclareActionSets(ctx, p.ID, p.Name, p.ActionSets); err != nil {
+		r.log.Warn("Declare plugin actionsets failed.", "pluginId", p.ID, "error", err)
```
Comment (marefr) [inline pkg/services/pluginsintegration/pipeline/steps.go:119]: Was the intent to log a warning and not return the error here? Otherwise would suggest using error level

Comment (eleijonmarck) [reply pkg/services/pluginsintegration/pipeline/steps.go:119]: most of the implementation is based off of roles registration. should we also log error there?

https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pipeline/steps.go#L91

File: pkg/plugins/models.go [hunk 7]
```diff
@@ -334,6 +334,17 @@ type Permission struct {
 	Scope  string `json:"scope"`
 }
 
+// ActionSetRegistration stores a actionset and its actions
+type ActionSetRegistration struct {
+	ActionSet ActionSet `json:"actionset"`
+}
+
+// ActionSet is the model for ActionSet in RBAC.
+type ActionSet struct {
```
Comment (aarongodin) [inline pkg/plugins/models.go:343]: I'm thinking about management operations related to action sets, and how that may impact updates or removals of plugin actionsets later.

Updating the actions of an existing actionset is straightfoward, since you have the name of the actionset to key off of. However changing the `action` is harder. At first I considered the `action` to be the unique identifier for the action set, but I'm worried that this isn't enough. If you want to rename an action set for any reason, there wouldn't be a way to do it without creating a new one, migrating all permissions already assigned to the new one (and those created in the future for some period of time), then eventually removing the old actionset. Otherwise, if you renamed one by doing a delete and create under the new name, you lose access already granted by the previous actionset.

As far as I know the only way around this is for the plugin author to specify a UID for each actionset. This is a bit of extra work but it could be a string derived from existing values on the plugin like it's ID.

Curious if you've already considered some of the day 2 operations that plugin authors might make with action sets?

Comment (eleijonmarck) [reply pkg/plugins/models.go:343]: @gamab i think me and gabriel talked about this. please correct me if i am wrong gabriel. We cannot hinder plugin authors too much about how to setup their own actionsets. We should absolutely support renaming a actionset.

The solutions might be for plugin authors, in order of ease:
- recommend to only build new actionsets and migrate to new once when permissions has been reassigned
- modify an existing actionset

> If you want to rename an action set for any reason, there wouldn't be a way to do it without creating a new one, migrating all permissions already assigned to the new one (and those created in the future for some period of time), then eventually removing the old actionset. 

honestly i think this is a viable strategy in my opinon. We do the same with some columns in databases. This might be the recommended approach from us and the same for us as well

**Review by gamab — APPROVED** [review]
LGTM! Great job to both of you!

As discussed with you, I think we should limit the action sets that can be modified to `folders:admin`, `folders:edit`, `folders:view`.

**Comment by gamab** [comment]
Side note: we will have to update the `plugin.json.schema` accordingly (and then plugin tools docs, and plugin example)

**Comment by gamab** [comment]
Post review question (better late than never :smile:): We don't write both action sets and their associated plugin permissions to the DB with this PR, is that intended?

**Comment by IevaVasiljeva** [comment]
> Post review question (better late than never 😄): We don't write both action sets and their associated plugin permissions to the DB with this PR, is that intended?

As in, we don't double-write (add an entry for the action set as well as entries for plugin permissions) to the DB? That is intended, I think it makes sense that plugin managed permissions only work if the feature toggle is enabled. Or do you think it's an issue?

**Comment by IevaVasiljeva** [comment]
@marefr and @wbrowne I'd appreciate a review of the updated PR when you have a bit of time 🙏

File: pkg/plugins/ifaces.go [hunk 8]
```diff
@@ -134,6 +134,11 @@ type RoleRegistry interface {
 	DeclarePluginRoles(ctx context.Context, ID, name string, registrations []RoleRegistration) error
 }
 
+// ActionSetRegistry handles the plugin RBAC actionsets
+type ActionSetRegistry interface {
```
Comment (wbrowne) [inline pkg/plugins/ifaces.go:138]: Unless you need to use this interface or any of its models within `pkg/plugins/*`, then I think this could actually be moved to `pluginsintegration` instead. If we only need to rely on this as a pipeline step (which looks to be the case), then we can consider it a "Grafana server" specific detail. 

Maybe the same can be said for `RoleRegistration`, but that's for another day 😄

**Review by wbrowne — COMMENTED** [review]
Code looks good AFAICT - just a matter of placement is all!

File: pkg/plugins/ifaces.go (hunk 8 above)
Comment (IevaVasiljeva) [reply pkg/plugins/ifaces.go:138]: Good point, I just stuck it next to `RoleRegistration`. I'll see if I can move both of them (if it's too big a change, I might do it in a separate PR).
